index.js

```diff
@@ -637,10 +666,47 @@ SendStream.prototype.sendFile = function sendFile(path) {
     fs.stat(p, function (err, stat) {
       if (err) return next(err)
       if (stat.isDirectory()) return next()
-      self.emit('file', p, stat)
-      self.send(p, stat)
+      checkPrecompressionAndSendFile(p, stat)
     })
   }
+
+  function checkPrecompressionAndSendFile(p, stat) {
+    self.emit('file', p, stat)
+    if (!self._precompressionFormats) return self.send(p, stat)
+
+    var state = {
+      contents: [],
+      extensionsToCheck: self._precompressionFormats.length
+    }
+
+    self._precompressionFormats.forEach(function (format) {
+      debug('stat "%s%s"', p, format.extension);
+      fs.stat(p + format.extension, function onstat(err, contentStat) {
```

Comment:
> Always sending Vary means that it is also sent for uncompressable files such as images. Which means that caching proxies have to keep about 5-10 (the number of unique accept-encoding headers used by common browsers) versions of each image in cache instead of only one.

Maybe a solution to this can be to allow the user to configure which types are actually compressed (similar to how the `compression` middleware works).

> If served files are changed while server is running then the browser/proxy caches will in any case have stale data and missing Vary header for a file that was cached before it had precompressed variants is correct behavior.

Sure, though by default, this module sends out `max-age=0` unless changed. Currently we have no issue with allowing files to change under this module, and there have been various issues relating to improving this over the years. As far as this module is concerned, any file can be changed at any time, this would include we should assume that a file that did not have a corresponding `.gz` file at one point in time may have one generated later, and should signal such in the response, i.m.o

> Maybe I (we?) should benchmark how much performance is lost by the extra stats?

Yes, please benchmark. Remember that Node.js absolutely sucks for any file operations, since all file operations (including stats) are not async, but rather executed sync in a background thread pool with a limited number of threads (defaults to 4). As such, influx in any file operations can really hurt Node.js-based applications.